### PR TITLE
Identity scaffolder

### PR DESCRIPTION
**Pull request by prafullbhosale**
#### Usage: 
```
dotnet aspnet-codegenerator identity -h


Usage: aspnet-codegenerator [arguments] [options]

Arguments:
  generator  Name of the generator. Check available generators below.

Options:
  -p|--project             Path to .csproj file in the project.
  -n|--nuget-package-dir
  -c|--configuration       Configuration for the project (Possible values: Debug/ Release)
  -tfm|--target-framework  Target Framework to use. (Short folder name of the tfm. eg. net46)
  -b|--build-base-path
  --no-build

Selected Code Generator: identity

Generator Options:
  --dbContext|-dc     : Name of the DbContext class to generate.
  --userClass|-u      : Name of the User class to generate.
  --useSqLite|-sqlite : Flag to specify if DbContext should use SQLite instead of SQL Server.
  --force|-f          : Use this option to overwrite existing files.

```

#### Generated content samples:
Without creating a custom user: https://github.com/prafullbhosale/identityscaffoldersample/pull/1
With creating a custom user: https://github.com/prafullbhosale/identityscaffoldersample/pull/2

cc @danroth27

### CODE REVIEW

File: src/VS.Web.CG.Mvc/Identity/IdentityGenerator.cs [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+{
+    [Alias("identity")]
+    public class IdentityGenerator : ICodeGenerator
+    {
+        private const string IdentityAreaName = "Identity";
+
+        private ILogger _logger;
+        private IApplicationInfo _applicationInfo;
+        private IServiceProvider _serviceProvider;
+        private ICodeGeneratorActionsService _codegeratorActionService;
```
Comment (seancpeters) [inline src/VS.Web.CG.Mvc/Identity/IdentityGenerator.cs]: Nit: spelling on variable name "gerator" -> "generator"

File: src/VS.Web.CG.Mvc/Identity/IdentityGenerator.cs [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+                await _codegeratorActionService.AddFileAsync(
+                    staticFile.Value,
+                    Path.Combine(TemplateFolderRoot, staticFile.Key)
+                );
+            }
+        }
+
+        private void Validate(IdentityGeneratorCommandLineModel model)
+        {
+            var errorStrings = new List<string>();;
```
Comment (seancpeters) [inline src/VS.Web.CG.Mvc/Identity/IdentityGenerator.cs]: Nit: ";;" in multiple places in this method

File: src/VS.Web.CG.Mvc/Identity/IdentityGenerator.cs [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+                    template.Value,
+                    template.Key,
+                    TemplateFolders,
+                    model);
+            }
+
+            var dbContextClass = model.IsGenerateCustomUser ? model.DbContextClass : "IdentityDbContext";
+            _connectionStringsWriter.AddConnectionString(
+                connectionStringName: $"{dbContextClass}Connection",
+                dataBaseName: $"{model.ApplicationName}",
```
Comment (seancpeters) [inline src/VS.Web.CG.Mvc/Identity/IdentityGenerator.cs]: Nit: no need for $ formatted string for dataBaseName

File: src/VS.Web.CG.Mvc/Templates/Identity/Pages/Account/Manage/Account.Manage.ExternalLogins.cs.cshtml [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+
+        [TempData]
+        public string StatusMessage { get; set; }
+
+        public async Task<IActionResult> OnGetAsync()
+        {
+            var user = await _userManager.GetUserAsync(User);
+            if (user == null)
+            {
+                return NotFound($"Unable to load user with ID '{_userManager.GetUserId(User)}'.");
```
[the remark this answers is not in the document]
Comment (seancpeters) [reply src/VS.Web.CG.Mvc/Templates/Identity/Pages/Account/Manage/Account.Manage.ExternalLogins.cs.cshtml]: This same pattern appears in a number of files, they should all probably be treated the same way as each other.

File: src/VS.Web.CG.Mvc/Templates/Identity/Pages/Error.cs.cshtml [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+    public class ErrorModel : PageModel
+    {
+        public string RequestId { get; set; }
+
+        public bool ShowRequestId => !string.IsNullOrEmpty(RequestId);
+
+        [ResponseCache(Duration = 0, Location = ResponseCacheLocation.None, NoStore = true)]
+        public void OnGet()
+        {
+            RequestId = Activity.Current?.Id ?? HttpContext.TraceIdentifier;
```
Comment (seancpeters) [inline src/VS.Web.CG.Mvc/Templates/Identity/Pages/Error.cs.cshtml]: If it's possible for Activity to be null, it should be null checked too, i.e.:
`Activity?.Current?.Id...`

File: src/VS.Web.CG.Mvc/Templates/Identity/Services/EmailSender.cshtml [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+
+namespace @(Model.Namespace).Areas.Identity.Services
+{
+    // This class is used by the application to send email for account confirmation and password reset.
+    // For more details see https://go.microsoft.com/fwlink/?LinkID=532713
+    public class EmailSender : IEmailSender
+    {
+        public Task SendEmailAsync(string email, string subject, string htmlMessage)
+        {
+            return Task.CompletedTask;
```
Comment (seancpeters) [inline src/VS.Web.CG.Mvc/Templates/Identity/Services/EmailSender.cshtml]: This method appears to be missing its functionality. Should this maybe be an abstract class or method?

File: src/VS.Web.CG.Mvc/Templates/Identity/Pages/Account/Manage/Account.Manage.ExternalLogins.cs.cshtml (hunk 4 above)
[the remark this answers is not in the document]
Comment (prafullbhosale) [reply src/VS.Web.CG.Mvc/Templates/Identity/Pages/Account/Manage/Account.Manage.ExternalLogins.cs.cshtml]: Not entirely sure of how the UserManager works, but the call to `GetUserId(User)` tries to find the identifier by the `User` instance claims and not by looking at a data store. @Javiercn/ @danroth27

File: src/VS.Web.CG.Mvc/Templates/Identity/Pages/Error.cs.cshtml (hunk 5 above)
Comment (prafullbhosale) [reply src/VS.Web.CG.Mvc/Templates/Identity/Pages/Error.cs.cshtml]: @javiercn / @danroth27, is `Activity` guaranteed to be not null over here?

File: src/VS.Web.CG.Mvc/Templates/Identity/Services/EmailSender.cshtml (hunk 6 above)
Comment (prafullbhosale) [reply src/VS.Web.CG.Mvc/Templates/Identity/Services/EmailSender.cshtml]: @javiercn / @danroth27 any thoughts?

**Comment by prafullbhosale** [comment]
Updated the PR to be against release/2.1 branch instead of dev

**Comment by prafullbhosale** [comment]
@javiercn @danroth27, I am going to merge this now. 
The Readme text changes are pending, but will get that in in a separate PR. 

cc @DamianEdwards , @mlorbetske

**Comment by javiercn** [comment]
This is a big too big to review in one sitting. Let's merge this in and we can tweak it later once we've used more if we need to.

**Pull request merged**